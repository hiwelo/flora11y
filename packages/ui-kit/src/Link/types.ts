import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react';

export type LinkProps = AnchorHTMLAttributes<HTMLAnchorElement> &
  ButtonHTMLAttributes<HTMLButtonElement>;
