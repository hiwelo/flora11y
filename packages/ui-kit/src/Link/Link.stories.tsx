import { withKnobs, text } from '@storybook/addon-knobs';
import React from 'react';

import Link from '.';

export default {
  component: Link,
  decorators: [withKnobs],
  title: 'Link',
};

export const overview: React.FC = () => (
  <Link href={text('Destination', '')}>{text('Content', 'Hello World!')}</Link>
);
