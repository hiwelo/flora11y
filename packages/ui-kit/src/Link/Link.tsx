import React, { FunctionComponent } from 'react';

import LinkElement from './components/LinkElement';
import { LinkProps } from './types';

const Link: FunctionComponent<LinkProps> = ({
  children,
  href,
  target,
  ...rest
}) => {
  const isLink = !!href;
  const computedRel = target === '_blank' ? 'noopener noreferrer' : undefined;
  const relAttr = isLink ? computedRel : undefined;
  const targetAttr = isLink ? target : undefined;

  return (
    <LinkElement href={href} rel={relAttr} target={targetAttr} {...rest}>
      {children}
    </LinkElement>
  );
};

export default Link;
