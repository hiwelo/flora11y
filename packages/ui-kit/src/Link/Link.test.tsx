/* eslint-disable jsx-a11y/anchor-is-valid */
import { render, screen } from '@testing-library/react';
import React from 'react';
import { create } from 'react-test-renderer';

import Link from '.';

const callback = jest.fn();

const testElements = (
  <>
    <Link data-testid="defaultLink" href="#test">
      Hello World!
    </Link>
    <Link data-testid="targetBlankLink" target="_blank" href="#test">
      Hello World!
    </Link>
    <Link
      data-testid="targetBlankLinkWithRelOverride"
      href="#test"
      rel={undefined}
      target="_blank"
    >
      Hello World!
    </Link>
    <Link
      data-testid="targetBlankLinkWithCustomRel"
      href="#test"
      rel="noopener"
      target="_blank"
    >
      Hello World!
    </Link>
    <Link data-testid="linkWithoutHref" onClick={callback}>
      Hello World!
    </Link>
  </>
);

describe('<Link />', () => {
  it('should adapt the markup based on the provided props', () => {
    expect.assertions(2);
    render(testElements);

    expect(screen.queryByTestId('defaultLink')?.tagName).toBe('A');
    expect(screen.queryByTestId('linkWithoutHref')?.tagName).toBe('BUTTON');
  });

  it('should add a rel="noopener noreferrer" to link with target="_blank"', () => {
    expect.assertions(6);
    render(testElements);

    const linkWithTargetBlank = screen.queryByTestId('targetBlankLink');
    const linkWithTargetBlankAndRelOverride = screen.queryByTestId(
      'targetBlankLinkWithRelOverride',
    );
    const linkWithTargetBlankAndCustomRel = screen.queryByTestId(
      'targetBlankLinkWithCustomRel',
    );

    // expect rel to be added by default
    expect(linkWithTargetBlank).toHaveAttribute('target', '_blank');
    expect(linkWithTargetBlank).toHaveAttribute('rel', 'noopener noreferrer');

    // expect rel to not be there if the addition was overwritten
    expect(linkWithTargetBlankAndRelOverride).toHaveAttribute(
      'target',
      '_blank',
    );
    expect(linkWithTargetBlankAndRelOverride).not.toHaveAttribute('rel');

    // expect rel to return a custom one if provided
    expect(linkWithTargetBlankAndCustomRel).toHaveAttribute('target', '_blank');
    expect(linkWithTargetBlankAndCustomRel).toHaveAttribute('rel', 'noopener');
  });

  it('should match snapshot', () => {
    expect.hasAssertions();
    const container = create(testElements);

    // eslint-disable-next-line jest/prefer-inline-snapshots
    expect(container.toJSON()).toMatchSnapshot();
  });
});
