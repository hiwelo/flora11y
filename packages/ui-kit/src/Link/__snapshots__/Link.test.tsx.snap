// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> should match snapshot 1`] = `
Array [
  <a
    className="sc-AxjAm gUSXVV"
    data-testid="defaultLink"
    href="#test"
  >
    Hello World!
  </a>,
  <a
    className="sc-AxjAm gUSXVV"
    data-testid="targetBlankLink"
    href="#test"
    rel="noopener noreferrer"
    target="_blank"
  >
    Hello World!
  </a>,
  <a
    className="sc-AxjAm gUSXVV"
    data-testid="targetBlankLinkWithRelOverride"
    href="#test"
    target="_blank"
  >
    Hello World!
  </a>,
  <a
    className="sc-AxjAm gUSXVV"
    data-testid="targetBlankLinkWithCustomRel"
    href="#test"
    rel="noopener"
    target="_blank"
  >
    Hello World!
  </a>,
  <button
    className="sc-AxjAm gUSXVV"
    data-testid="linkWithoutHref"
    onClick={[MockFunction]}
    type="button"
  >
    Hello World!
  </button>,
]
`;
